tests/publish-dsl.nf

```diff
@@ -0,0 +1,85 @@
+#!/usr/bin/env nextflow
+/*
+ * Copyright 2013-2024, Seqera Labs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+nextflow.preview.topic = true
+
+process align {
+  input:
+  val(x)
+
+  output:
+  path("*.bam")
+  path("${x}.bai")
+
+  """
+  echo ${x} > ${x}.bam
+  echo ${x} | rev > ${x}.bai
+  """
+}
+
+process my_combine {
+  input:
+  path(bamfile)
+  path(baifile)
+
+  output:
+  path 'result.txt'
+
+  """
+  cat $bamfile > result.txt
+  cat $baifile >> result.txt
+  """
+}
+
+process foo {
+  output:
+  path 'xxx'
+
+  '''
+  mkdir xxx
+  touch xxx/A
+  touch xxx/B
+  touch xxx/C
+  '''
+}
+
+workflow {
+  def input = Channel.of('alpha','beta','delta')
+  align(input)
+
+  def bam = align.out[0].toSortedList { it.name }
+  def bai = align.out[1].toSortedList { it.name }
+  my_combine( bam, bai )
+  my_combine.out.view{ it.text }
+
+  foo | topic('foo')
+}
+
+output {
+  directory 'results'
+
+  'data' {
+    select align.out[0], mode: 'copy'
```

Comment:
Also not sure we should use `select` here, we could use `from` to avoid introducing another keyword

Reply:
All publish options are supported at any level. The example you proposed can be written using the `defaults` directive:
```groovy
output {
  directory 'result'
  defaults mode: "symlink"

  'data' {
    defaults mode: 'copy'
    select align.out[0]
  }
}
```

This works for every publish option, not just mode, so that users can transition from publishDir to the output DSL without losing functionality.

> Also not sure we should use select here, we could use from to avoid introducing another keyword

I'm not married to a particular keyword, `from` seems fine to me. Would you then use `fromTopic` for the topic selector?

Reply:
> The example you proposed can be written using the defaults directive

Not aware of that. From where does it come from? 


> from seems fine to me. Would you then use fromTopic for the topic selector?

Ideally, always `from` could be used

Reply:
Sounds good, string -> from topic and channel -> from channel

Check out the docs I wrote in this PR (or the docs deploy preview), everything is explained there

Reply:
We touched on this briefly while meeting yesterday, but cannot recall where we landed: for the sake of more concise code, it would be good if the keyword `defaults` could be avoided

Reply:
I have stuck with `defaults` instead of explicit methods for each publish option so that the syntax for publish options is the same for directory defaults and individual `from` selectors. Otherwise I think it would be harder to understand:
```groovy
'results' {
  mode 'copy'
  'foo' {
    from foo.out, pattern: '*.txt'
  }
}
```

It's not as clear here that `mode` and `pattern` are both publish options being specified with different syntax

Reply:
Agree on the non clarity of yout snippet above. What prevents us from adopting the following? 
```groovy
'results' {
  mode 'copy'
  'foo' {
    pattern: '*.txt'
    from foo.out
  }
}
```

Reply:
Because then the pattern applies to the entire directory instead of the individual rule. These options often need to be applied to individual rules, so if they can only be specified for a directory and there are multiple rules in the same directory, you would have to repeat the directory which is burdensome:
```groovy
'results' {
  mode 'copy'
  'foo' {
    pattern '*.txt'
    from foo.out
  }
  'foo' {
    from foobar.out
  }

  // instead of this
  'foo' {
    from foo.out, pattern: '*.txt'
    from foobar.out
  }
}
```

Reply:
This dsl should be an opportunity for simplification. I'm with Marco that options should be at group level not at from level. It's very likely that the same pattern needs to be applied to all `from` statements

Reply:
> It's very likely that the same pattern needs to be applied to all from statements

If you look at the rnaseq prototype you will see that this is simply not true

I am all for simplification but we cannot simplify to the point of not supporting the behavior that is needed

Reply:
The output DSL has already simplified things in three major ways:
1. the publish mode is specified once for the entire directory instead of once for every rule
2. putting the directory structure first, instead of specifying the full path for every rule, removes a lot of duplication of base paths
3. the pattern option is often not needed because you can select individual channels instead of entire processes

Reply:
Then you need to demonstrate the bad pattern and offer something better. Right now you are hand waving and not really engaging with the actual pipeline

Reply:
Thanks for clarifying and for all the feedbacks.   

I now better appreciate the merits of enabling both path-level and rule-specific options, considering some big pipelines. 

I think I am a bit distracted by `defaults` both adding verbosity and at the same time not being super expressive about the meaning.

In other words, I would not find this confusing (open to other views, incl scidev):

```groovy
'results' {
  mode 'copy'
  'foo' {
    mode 'symlink'
    from foo.out, pattern: '*.txt'
  }
}
```

As long as we document that all publish options can be applied at both path level (standalone line) and rule level (comma separated within a `from` specification). 

There might be some degree of error proneness (spare or missing comma may trigger unwanted results), but we could promote the best practice of always having the path level options stated **before** the list of `from` specifications.